### ISSUE TITLE
2 Crashes due to the ML Kit module not getting installed properly

### ISSUE THREAD

**Issue by xrigau** [onfido/onfido-android-sdk#212]
#### What was the expected behaviour?

The app shouldn't crash, the Onfido flow should work just fine. If there's an error initialising the MLKit library, the Onfido SDK should catch the error and handle it gracefully (show an error, or use the camera without the image recognition capabilities).

#### What happened instead?

The app crashes while going through the Onfido screens.

#### Version info:

- `onfido-capture-sdk` version: `com.onfido.sdk.capture:onfido-capture-sdk-core:12.3.0` - Notice we're using `onfido-capture-sdk-core`
- Android API version(s): According to Firebase, this crash has occurred in Android 10, 11 and 12.

- `compileSdkVersion`: 33
- `targetSdkVersion`: 31
- `minSdkVersion`: 27
- Support library version (if applicable): N/A

#### Steps to reproduce:

We don't know, but it seems as if some devices are unable to load/install the MLKit library that the Onfido SDK uses, resulting in a crash during the Onfido ID verification flow.

#### Onfido Flow initialization code:

```kotlin
val tokenExpirationHandler = // ...
    val config = OnfidoConfig.builder(context)
        .withSDKToken("token", tokenExpirationHandler)
        .build()
    val intent = onfido.createIntent(config)
    startActivity(intent)
```

#### Issue severity information (number of users/crashes if applicable):

The 2 crashes below have occurred a total of 325 times and affect 43 users for the past couple of weeks.

#### Additional LogCat output:

Stacktrace from Firebase:

```
Fatal Exception: io.reactivex.rxjava3.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | com.google.mlkit.common.MlKitException: Waiting for the text optional module to be downloaded. Please wait.
       at io.reactivex.rxjava3.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:372)
       at io.reactivex.rxjava3.subjects.PublishSubject$PublishDisposable.onError(PublishSubject.java:316)
       at io.reactivex.rxjava3.subjects.PublishSubject.onError(PublishSubject.java:241)
       at io.reactivex.rxjava3.internal.operators.observable.ObservablePublishSelector$SourceObserver.onError(ObservablePublishSelector.java:85)
       at io.reactivex.rxjava3.internal.observers.BasicFuseableObserver.onError(BasicFuseableObserver.java:100)
       at io.reactivex.rxjava3.internal.observers.BasicFuseableObserver.onError(BasicFuseableObserver.java:100)
       at io.reactivex.rxjava3.internal.operators.observable.ObservableScan$ScanObserver.onError(ObservableScan.java:104)
       at io.reactivex.rxjava3.internal.observers.BasicFuseableObserver.onError(BasicFuseableObserver.java:100)
       at io.reactivex.rxjava3.internal.observers.BasicFuseableObserver.onError(BasicFuseableObserver.java:100)
       at io.reactivex.rxjava3.observers.SerializedObserver.onError(SerializedObserver.java:156)
       at io.reactivex.rxjava3.internal.operators.observable.ObservableSkipUntil$SkipUntilObserver.onError(ObservableSkipUntil.java:81)
       at io.reactivex.rxjava3.subjects.PublishSubject$PublishDisposable.onError(PublishSubject.java:318)
       at io.reactivex.rxjava3.subjects.PublishSubject.onError(PublishSubject.java:241)
       at io.reactivex.rxjava3.internal.operators.observable.ObservablePublishSelector$SourceObserver.onError(ObservablePublishSelector.java:85)
       at io.reactivex.rxjava3.internal.observers.BasicFuseableObserver.onError(BasicFuseableObserver.java:100)
       at io.reactivex.rxjava3.internal.operators.observable.ObservableRefCount$RefCountObserver.onError(ObservableRefCount.java:207)
       at io.reactivex.rxjava3.internal.operators.observable.ObservablePublish$PublishConnection.onError(ObservablePublish.java:191)
       at io.reactivex.rxjava3.internal.observers.BasicFuseableObserver.onError(BasicFuseableObserver.java:100)
       at io.reactivex.rxjava3.internal.util.AtomicThrowable.tryTerminateConsumer(AtomicThrowable.java:110)
       at io.reactivex.rxjava3.internal.operators.observable.ObservableFlatMapSingle$FlatMapSingleObserver.drainLoop(ObservableFlatMapSingle.java:228)
       at io.reactivex.rxjava3.internal.operators.observable.ObservableFlatMapSingle$FlatMapSingleObserver.drain(ObservableFlatMapSingle.java:200)
       at io.reactivex.rxjava3.internal.operators.observable.ObservableFlatMapSingle$FlatMapSingleObserver.innerError(ObservableFlatMapSingle.java:194)
       at io.reactivex.rxjava3.internal.operators.observable.ObservableFlatMapSingle$FlatMapSingleObserver$InnerObserver.onError(ObservableFlatMapSingle.java:273)
       at io.reactivex.rxjava3.internal.operators.single.SingleDoFinally$DoFinallyObserver.onError(SingleDoFinally.java:79)
       at io.reactivex.rxjava3.internal.operators.single.SingleFlatMap$SingleFlatMapCallback$FlatMapSingleObserver.onError(SingleFlatMap.java:117)
       at io.reactivex.rxjava3.internal.operators.single.SingleMap$MapSingleObserver.onError(SingleMap.java:70)
       at io.reactivex.rxjava3.internal.operators.single.SingleCreate$Emitter.tryOnError(SingleCreate.java:95)
       at io.reactivex.rxjava3.internal.operators.single.SingleCreate$Emitter.onError(SingleCreate.java:81)
       at com.onfido.android.sdk.capture.internal.util.TaskExtensionKt.toSingle$lambda-2$lambda-1(TaskExtensionKt.java:11)
       at com.onfido.android.sdk.capture.internal.util.TaskExtensionKt.$r8$lambda$No8IH6QdMeNqHWiD4BvqdGrojrU(TaskExtensionKt.java)
       at com.onfido.android.sdk.capture.detector.mrz.MRZDetectorGoogle$$InternalSyntheticLambda$0$f2a26d097950a716c90487c0b1b06b066d898e5e79d6a5589fda1c2203d4ee5a$1.onFailure$bridge(MRZDetectorGoogle.java:14)
       at com.google.android.gms.tasks.zzk.run(zzk.java:1)
       at com.google.android.gms.common.api.internal.zacr.run$bridge(com.google.android.gms:play-services-base@@18.0.1:28)
       at io.reactivex.rxjava3.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:38)
       at io.reactivex.rxjava3.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:25)
       at java.util.concurrent.FutureTask.run(FutureTask.java:264)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:307)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
       at java.lang.Thread.run(Thread.java:1012)

Caused by com.google.mlkit.common.MlKitException: Waiting for the text optional module to be downloaded. Please wait.
       at com.google.mlkit.vision.text.internal.zzb.zzb(com.google.android.gms:play-services-mlkit-text-recognition-common@@17.0.0:16)
       at com.google.mlkit.vision.text.internal.TextRecognizerTaskWithResource.load(com.google.android.gms:play-services-mlkit-text-recognition-common@@17.0.0:1)
       at com.google.mlkit.common.sdkinternal.ModelResource.zza(ModelResource.java:4)
       at com.google.mlkit.common.sdkinternal.zzl.run(com.google.mlkit:common@@18.1.0:4)
       at com.google.mlkit.common.sdkinternal.zzp.run(com.google.mlkit:common@@18.1.0:2)
       at com.google.mlkit.common.sdkinternal.MlKitThreadPool.zze(com.google.mlkit:common@@18.1.0:4)
       at com.google.mlkit.common.sdkinternal.zzi.run(com.google.mlkit:common@@18.1.0:10)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
       at com.google.mlkit.common.sdkinternal.MlKitThreadPool.zzd(MlKitThreadPool.java:2)
       at com.google.mlkit.common.sdkinternal.zzj.run(com.google.mlkit:common@@18.1.0:2)
       at java.lang.Thread.run(Thread.java:1012)
```

And then we also have this other crash in Firebase, which we think is the same or very closely related:

```
Fatal Exception: io.reactivex.rxjava3.exceptions.OnErrorNotImplementedException: The exception was not handled due to missing onError handler in the subscribe() method call. Further reading: https://github.com/ReactiveX/RxJava/wiki/Error-Handling | com.google.mlkit.common.MlKitException: Waiting for the text optional module to be downloaded. Please wait.
       at io.reactivex.rxjava3.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:718)
       at io.reactivex.rxjava3.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:715)
       at io.reactivex.rxjava3.internal.observers.LambdaObserver.onError(LambdaObserver.java:77)
       at io.reactivex.rxjava3.internal.observers.DisposableLambdaObserver.onError(DisposableLambdaObserver.java:65)
       at io.reactivex.rxjava3.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.checkTerminated(ObservableObserveOn.java:281)
       at io.reactivex.rxjava3.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:172)
       at io.reactivex.rxjava3.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:255)
       at io.reactivex.rxjava3.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:123)
       at android.os.Handler.handleCallback(Handler.java:938)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loopOnce(Looper.java:226)
       at android.os.Looper.loop(Looper.java:313)
       at android.app.ActivityThread.main(ActivityThread.java:8663)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:571)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1135)

Caused by com.google.mlkit.common.MlKitException: Waiting for the text optional module to be downloaded. Please wait.
       at com.google.mlkit.vision.text.internal.zzb.zzb(com.google.android.gms:play-services-mlkit-text-recognition-common@@17.0.0:16)
       at com.google.mlkit.vision.text.internal.TextRecognizerTaskWithResource.load(com.google.android.gms:play-services-mlkit-text-recognition-common@@17.0.0:1)
       at com.google.mlkit.common.sdkinternal.ModelResource.zza(ModelResource.java:4)
       at com.google.mlkit.common.sdkinternal.zzl.run(com.google.mlkit:common@@18.1.0:4)
       at com.google.mlkit.common.sdkinternal.zzp.run(com.google.mlkit:common@@18.1.0:2)
       at com.google.mlkit.common.sdkinternal.MlKitThreadPool.zze(com.google.mlkit:common@@18.1.0:4)
       at com.google.mlkit.common.sdkinternal.zzi.run(com.google.mlkit:common@@18.1.0:10)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
       at com.google.mlkit.common.sdkinternal.MlKitThreadPool.zzd(MlKitThreadPool.java:2)
       at com.google.mlkit.common.sdkinternal.zzj.run(com.google.mlkit:common@@18.1.0:2)
       at java.lang.Thread.run(Thread.java:1012)
```

#### Other useful info:

Does not seem to be a device-specific issue, as it occurs on Samsung, Huawei, Tecno and Safaricom devices, amongst others.
I tried to do some research online but couldn't find anything helpful, in my understanding, this library should get automatically installed, but if that installation fails, we should handle the crash in a better way.

https://developers.google.com/ml-kit/vision/text-recognition/android
![image](https://user-images.githubusercontent.com/1626673/194591783-15d8e61a-5769-4fdd-9d2f-b1fea5ada1d7.png)

**Comment by gurpreet-onfido** [comment]
Hello @xrigau 

Thank you for bringing this up. 

We are checking this internally with the SDK team and we will get back to you on this as soon as possible. 

Kind regards
Gurpreet Singh

**Comment by xrigau** [comment]
Great, thank you @gurpreet-onfido !

**Comment by xrigau** [comment]
@gurpreet-onfido this is affecting 0.8% of our checks (134 crashes out of 17k checks for the past 7 days), it's a small amount compared to the total number of checks, but this is our top 2 crash in Firebase at the moment

**Comment by leewapadmi** [comment]
This has also happened on our app. Interestingly, it has only happened for users who are based outside of the United Kingdom. I don't know if that helps. It seems to work fine on an emulator.

**Comment by rovkinmax** [comment]
I have the same issue with one of our user
Device: Samsung Galaxy Z Fold3 5G
Operating System: Android 12

**Comment by gurpreet-onfido** [comment]
Hi there!

Greetings for the day!

I am glad to update you that the reported issue is now resolved. We have released a new version ([13.0.1](https://github.com/onfido/onfido-android-sdk/releases/tag/13.0.1)) of our SDK that contains the fix for this issue. 

You can find the release notes here: https://github.com/onfido/onfido-android-sdk/releases

Kindly upgrade to this release and let us know if this issue still persists. 

Kind regards
Gurpreet Singh

**Comment by xrigau** [comment]
Awesome, thanks for the update @gurpreet-onfido !
